util/timer.h

```diff
@@ -151,14 +155,34 @@ class Timer {
   }
 
   void CancelAllWithLock() {
+    mutex_.AssertHeld();
     if (map_.empty() && heap_.empty()) {
       return;
     }
 
+    // With mutex_ held, set all tasks to invalid so that they will not be
+    // re-queued.
+    // TODO: maybe use a single flag to prevent re-queuing.
+    for (auto& elem : map_) {
+      auto& func_info = elem.second;
+      assert(func_info);
+      func_info->valid = false;
+    }
+
+    if (!heap_.empty()) {
+      const FunctionInfo& top = *(heap_.top());
+      // WaitForTaskCompleteIfNecessary() may release mutex_
+      WaitForTaskCompleteIfNecessary(top);
+    }
+
     while (!heap_.empty()) {
+#ifndef NDEBUG
+      const FunctionInfo* func_info = heap_.top();
+      assert(func_info);
+      assert(!func_info->IsExecuting());
+#endif  // !NDEBUG
       heap_.pop();
     }
-
     map_.clear();
   }
```

Comment:
Not sure if it's a valid usecase: what if there're 2 threads trying to shutdown at the same time? For example, 1st thread try to shutdown, which may wait in `WaitForTaskCompleteIfNecessary()`, then 2nd thread try to call `shutdown` too, which stops in `WaitForTaskCompleteIfNecessary()` too if the function is still running. One of them will awake first and clean up all functions. Then the other will check invalid `fn.executing` when it's awaken eventually.

Reply:
Will update.